mesonbuild/scripts/depaccumulate.py

```diff
@@ -0,0 +1,130 @@
+
+
+def quote(text: str) -> str:
+    # Fast path for when no quoting is necessary
+    if not _QUOTE_PAT.search(text):
+        return text
+    if '\n' in text:
+        errmsg = textwrap.dedent(f'''\
+            Ninja does not support newlines in rules. The content was:
+
+            {text}
+
+            Please report this error with a test case to the Meson bug tracker.''')
+        raise RuntimeError(errmsg)
+    return _QUOTE_PAT.sub(r'$\g<0>', text)
+
+
+_PROVIDER_CACHE: T.Dict[str, str] = {}
+
+
+def get_provider(rules: T.List[Rule], name: str) -> T.Optional[str]:
+    """Get the object that a module from another Target provides
+
+    We must rely on the object file here instead of the module itself, because
+    the object rule is part of the generated build.ninja, while the module is
+    only declared inside a dyndep. This creates for the dyndep generator to
+    depend on previous dyndeps as order deps. Since the module
+    interface file will be generated when the object is generated we can rely on
+    that in proxy and simplify generation.
+
+    :param rules: The list of rules to check
+    :param name: The logical-name to look for
+    :raises RuntimeError: If no provider can be found
+    :return: The object file of the rule providing the module
+    """
+    # Cache the result for performance reasons
+    if name in _PROVIDER_CACHE:
+        return _PROVIDER_CACHE[name]
+
+    for r in rules:
+        for p in r.get('provides', []):
+            if p['logical-name'] == name:
+                obj = r['primary-output']
+                _PROVIDER_CACHE[name] = obj
+                return obj
+    return None
+
+
+def process_rules(rules: T.List[Rule],
+                  extra_rules: T.List[Rule],
+                  ) -> T.Iterable[T.Tuple[str, T.Optional[T.List[str]], T.Optional[T.List[str]]]]:
+    """Process the rules for this Target
+
+    :param rules: the rules for this target
+    :param extra_rules: the rules for all of the targets this one links with, to use their provides
+    :yield: A tuple of the output, the exported modules, and the consumed modules
+    """
+    for rule in rules:
+        prov: T.Optional[T.List[str]] = None
+        req: T.Optional[T.List[str]] = None
+        if 'provides' in rule:
+            prov = [p['compiled-module-path'] for p in rule['provides']]
+        if 'requires' in rule:
+            req = []
+            for p in rule['requires']:
+                modfile = p.get('compiled-module-path')
+                if modfile is not None:
+                    req.append(modfile)
+                else:
+                    # We can't error if this is not found because of compiler
+                    # provided modules
+                    found = get_provider(extra_rules, p['logical-name'])
+                    if found:
+                        req.append(found)
+        yield rule['primary-output'], prov, req
+
+
+def format(files: T.Optional[T.List[str]]) -> str:
```

Comment:
This function shadows the builtin `format` function...

Reply:
I didn't even know there was a builtin format function...